src/librawspeed/decoders/Cr2Decoder.cpp

```diff
@@ -457,236 +445,93 @@ void Cr2Decoder::interpolate_420(int w, int h, int start_h , int end_h) {
+      store(n_line, r, g, b, off + 3, off + 4, off + 5);
       off += 6;
     }
   }
 }
 
-#undef YUV_TO_RGB
-
-#define YUV_TO_RGB(Y, Cb, Cr)                                                  \
-  r = sraw_coeffs[0] * ((Y) + (Cr)-512);                                       \
-  g = sraw_coeffs[1] * ((Y) + ((-778 * (Cb) - ((Cr) << 11)) >> 12) - 512);     \
-  b = sraw_coeffs[2] * ((Y) + ((Cb)-512));                                     \
-  r >>= 8;                                                                     \
-  g >>= 8;                                                                     \
+void Cr2Decoder::YUV_TO_RGB_v0(int Y, int Cb, int Cr, int& r, int& g, int& b,
+                               const int* sraw_coeffs) {
+  r = sraw_coeffs[0] * (Y + ((50 * Cb + 22929 * Cr) >> 12));
+  g = sraw_coeffs[1] * (Y + ((-5640 * Cb - 11751 * Cr) >> 12));
+  b = sraw_coeffs[2] * (Y + ((29040 * Cb - 101 * Cr) >> 12));
+  r >>= 8;
+  g >>= 8;
   b >>= 8;
+}
 
-// Note: Thread safe.
-
-void Cr2Decoder::interpolate_422_old(int w, int h, int start_h , int end_h) {
-  // Last pixel should not be interpolated
-  w--;
+static void STORE_RGB(ushort16* X, int r, int g, int b, int A, int B, int C) {
+  X[A] = clampbits(r, 16);
+  X[B] = clampbits(g, 16);
+  X[C] = clampbits(b, 16);
+}
 
-  // Current line
-  ushort16* c_line;
-  const int hue = -getHue() + 16384;
+void Cr2Decoder::interpolate_422(int w, int h, int start_h, int end_h) {
+  auto hue = -getHue() + 16384;
+  _interpolate_422(YUV_TO_RGB_v0, STORE_RGB, sraw_coeffs, mRaw, hue, hue, w, h,
+                   start_h, end_h);
+}
 
-  for (int y = start_h; y < end_h; y++) {
-    c_line = (ushort16*)mRaw->getData(0, y);
-    int r, g, b;
-    int off = 0;
-    for (int x = 0; x < w; x++) {
-      int Y = c_line[off];
-      int Cb = c_line[off+1] - hue;
-      int Cr = c_line[off+2] - hue;
-      YUV_TO_RGB(Y, Cb, Cr);
-      STORE_RGB(c_line, off, off + 1, off + 2);
-      off += 3;
+void Cr2Decoder::interpolate_420(int w, int h, int start_h, int end_h) {
+  auto hue = -getHue() + 16384;
+  _interpolate_420(YUV_TO_RGB_v0, STORE_RGB, sraw_coeffs, mRaw, hue, w, h,
+                   start_h, end_h);
+}
 
-      Y = c_line[off];
-      int Cb2 = (Cb + c_line[off+1+3] - hue) >> 1;
-      int Cr2 = (Cr + c_line[off+2+3] - hue) >> 1;
-      YUV_TO_RGB(Y, Cb2, Cr2);
-      STORE_RGB(c_line, off, off + 1, off + 2);
-      off += 3;
-    }
-    // Last two pixels
-    int Y = c_line[off];
-    int Cb = c_line[off+1] - 16384;
-    int Cr = c_line[off+2] - 16384;
-    YUV_TO_RGB(Y, Cb, Cr);
-    STORE_RGB(c_line, off, off + 1, off + 2);
+/* Algorithm found in EOS 40D */
+void Cr2Decoder::YUV_TO_RGB_v1(int Y, int Cb, int Cr, int& r, int& g, int& b,
+                               const int* sraw_coeffs) {
+  r = sraw_coeffs[0] * (Y + Cr - 512);
+  g = sraw_coeffs[1] * (Y + ((-778 * Cb - (Cr << 11)) >> 12) - 512);
+  b = sraw_coeffs[2] * (Y + (Cb - 512));
+  r >>= 8;
+  g >>= 8;
+  b >>= 8;
+}
```

Comment:
i did inline `store`, but the `A`, `B` and `C` need to be passed. or even more of that function needs to be rewritten. they should probably be rewritten even more, but not now :)